src/main/java/uk/gov/hmcts/divorce/bulkaction/ccd/event/CaseworkerScheduleCase.java

```diff
@@ -119,4 +139,176 @@ public SubmittedCallbackResponse submitted(CaseDetails<BulkActionCaseData, BulkA
+        }
+
+        for (Map.Entry<String, Integer> entry : idFrequencyMap.entrySet()) {
+            if (entry.getValue() > 1) {
+                if (!duplicateCaseIds.contains(entry.getKey())) {
+                    duplicateCaseIds.add(entry.getKey());
+                }
+            }
+        }
+        uniqueIds.removeAll(duplicateCaseIds);
+
+        return new DuplicateCheckResult(duplicateCaseIds, uniqueIds);
+    }
+
+    private List<String> getMissingIds(
+        final List<uk.gov.hmcts.reform.ccd.client.model.CaseDetails> searchResultsList,
+        final List<String> searchedIds
+    ) {
+        if (searchResultsList.size() < searchedIds.size()) {
+            final List<String> foundIds = searchResultsList.stream().map(caseDetails -> caseDetails.getId().toString()).toList();
+            return searchedIds.stream().filter(id -> !foundIds.contains(id)).toList();
+        }
+        return new ArrayList<>();
+    }
+
+    private DuplicateCheckResult getNewCaseIds(final List<ListValue<BulkListCaseDetails>> bulkListCaseDetails,
+                                       final List<ListValue<BulkListCaseDetails>> beforeBulkListCaseDetails
+    ) {
+        final DuplicateCheckResult duplicateCheckResult = checkForDuplicates(bulkListCaseDetails);
+        if (beforeBulkListCaseDetails == null) {
+            return duplicateCheckResult;
+        }
+
+        final DuplicateCheckResult filteredDuplicateCheckResult = checkForDuplicates(bulkListCaseDetails.stream()
+            .filter(caseDetails -> !beforeBulkListCaseDetails.contains(caseDetails)).toList());
+
+        if (filteredDuplicateCheckResult.hasUniqueIds()) {
+            final List<String> duplicateIdsToRemove = new ArrayList<>();
+            for (String id : filteredDuplicateCheckResult.uniqueIds()) {
+                if (duplicateCheckResult.duplicateIds().contains(id)) {
+                    duplicateIdsToRemove.add(id);
+                }
+            }
+            filteredDuplicateCheckResult.uniqueIds().removeAll(duplicateIdsToRemove);
+        }
+
+        return filteredDuplicateCheckResult;
+    }
+
+    private List<String> checkForRemovedCases(
+        final List<ListValue<BulkListCaseDetails>> bulkListCaseDetails,
+        final List<ListValue<BulkListCaseDetails>> beforeBulkListCaseDetails
+    ) {
+        if (beforeBulkListCaseDetails != null) {
+            final List<String> beforeCaseRefs = beforeBulkListCaseDetails.stream()
+                .map(caseDetails -> caseDetails.getValue().getCaseReference().getCaseReference())
+                .toList();
+            final List<String> caseRefs = bulkListCaseDetails.stream()
+                .map(caseDetails -> caseDetails.getValue().getCaseReference().getCaseReference())
+                .toList();
+            return beforeCaseRefs.stream()
+                .filter(beforeCaseRef -> !caseRefs.contains(beforeCaseRef))
+                .toList();
+        }
```

Comment:
We could use guard clauses instead of nested ifs to make some of the new methods more readable.

We could also reduce the number of times we loop through this and use a set for O(1) rather than O(N) lookups. They're only small collections, but good to be in the habit of using a set where appropriate.

```suggestion
        if (beforeBulkListCaseDetails == null) {
          return Collections.emptyList();
        }
        
         final List<String> caseRefs = bulkListCaseDetails.stream()
                .map(caseDetails -> caseDetails.getValue().getCaseReference().getCaseReference())
                .collect(Collectors.toSet());

            return beforeBulkListCaseDetails.stream()
                .map(caseDetails -> caseDetails.getValue().getCaseReference().getCaseReference())
                .filter(beforeRef -> !caseRefs.contains(beforeRef))
                .toList();
        }
```